tls/s2n_prf.c

```diff
@@ -78,38 +80,208 @@ static int s2n_sslv3_prf(union s2n_prf_working_space *ws, struct s2n_blob *secre
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_P_HASH_INVALID_ALGORITHM);
+    }
+
+    /* Initialize the mac key using the provided secret */
+    notnull_check(ws->tls.p_hash.evp.mac_key = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, secret->data, secret->size));
+
+    /* Initialize the message digest context with the above message digest and mac key */
+    return s2n_evp_p_hash_digest_init(ws);
+}
+
+static int s2n_evp_p_hash_update(union s2n_prf_working_space *ws, const void *data, uint32_t size)
+{
+    if (EVP_DigestSignUpdate(ws->tls.p_hash.evp.md_ctx, data, (size_t)size) == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_UPDATE_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_evp_p_hash_digest(union s2n_prf_working_space *ws, void *digest, uint32_t size)
+{
+    /* EVP_DigestSign API's require size_t data structures */
+    size_t digest_size = size;
+
+    if (EVP_DigestSignFinal(ws->tls.p_hash.evp.md_ctx, (unsigned char *)digest, &digest_size) == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_FINAL_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_evp_p_hash_wipe(union s2n_prf_working_space *ws)
+{
+    if (S2N_EVP_MD_CTX_RESET(ws->tls.p_hash.evp.md_ctx) == 0) {
+        S2N_ERROR(S2N_ERR_P_HASH_WIPE_FAILED);
+    }
+
+    return 0;
+}
+
+static int s2n_evp_p_hash_reset(union s2n_prf_working_space *ws)
+{
+    GUARD(s2n_evp_p_hash_wipe(ws));
+
+    return s2n_evp_p_hash_digest_init(ws);
+}
+
+static int s2n_evp_p_hash_cleanup(union s2n_prf_working_space *ws)
+{
+    /* Prepare the workspace md_ctx for the next p_hash */
+    GUARD(s2n_evp_p_hash_wipe(ws));
+
+    /* Free mac key - PKEYs cannot be reused */
+    notnull_check(ws->tls.p_hash.evp.mac_key);
+    EVP_PKEY_free(ws->tls.p_hash.evp.mac_key);
+    ws->tls.p_hash.evp.mac_key = NULL;
+
+    return 0;
+}
+
+static int s2n_evp_p_hash_free(union s2n_prf_working_space *ws)
+{
+    notnull_check(ws->tls.p_hash.evp.md_ctx);
+    S2N_EVP_MD_CTX_FREE(ws->tls.p_hash.evp.md_ctx);
+    ws->tls.p_hash.evp.md_ctx = NULL;
+
+    return 0;
+}
+
+static int s2n_hmac_p_hash_new(union s2n_prf_working_space *ws)
+{
+    return 0;
+}
+
+static int s2n_hmac_p_hash_init(union s2n_prf_working_space *ws, s2n_hmac_algorithm alg, struct s2n_blob *secret)
+{
+    return s2n_hmac_init(&ws->tls.p_hash.hmac, alg, secret->data, secret->size);
+}
+
+static int s2n_hmac_p_hash_update(union s2n_prf_working_space *ws, const void *data, uint32_t size)
+{
+    return s2n_hmac_update(&ws->tls.p_hash.hmac, data, size);
+}
+
+static int s2n_hmac_p_hash_digest(union s2n_prf_working_space *ws, void *digest, uint32_t size)
+{
+    return s2n_hmac_digest(&ws->tls.p_hash.hmac, digest, size);
+}
+
+static int s2n_hmac_p_hash_reset(union s2n_prf_working_space *ws)
+{
+    return s2n_hmac_reset(&ws->tls.p_hash.hmac);
+}
+
+static int s2n_hmac_p_hash_cleanup(union s2n_prf_working_space *ws)
+{
+    return 0;
+}
+
+static int s2n_hmac_p_hash_free(union s2n_prf_working_space *ws)
+{
+    return 0;
+}
+
```

Comment:
The source of truth on why we have two implementations is in s2n_prf_init(). If s2n_prf_init changes, these comments can easily become stale. I'm about to upload a new revision with a little more substance to the s2n_prf_init comment and we can discuss this further if you would still prefer additional comments here.